src/Renci.SshNet/SftpClient.cs

```diff
@@ -809,6 +958,39 @@ public void UploadFile(Stream input, string path, bool canOverride, Action<ulong
             InternalUploadFile(input, path, flags, null, uploadCallback);
         }
 
+#if FEATURE_TAP
+        /// <summary>
+        /// Asynchronously uploads stream into remote file.
+        /// </summary>
+        /// <param name="input">Data input stream.</param>
+        /// <param name="path">Remote file path.</param>
+        /// <param name="createMode">Specifies how the file should be created.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous upload operation.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="input" /> is <b>null</b> -or- <paramref name="path" /> is <b>null</b>.</exception>
+        /// <exception cref="SshConnectionException">Client is not connected.</exception>
+        /// <exception cref="SftpPermissionDeniedException">Permission to upload the file was denied by the remote host. <para>-or-</para> A SSH command was denied by the server.</exception>
+        /// <exception cref="SshException">A SSH error where <see cref="Exception.Message" /> is the message from the remote host.</exception>
+        /// <exception cref="ObjectDisposedException">The method was called after the client was disposed.</exception>
+        /// <remarks>
+        /// Method calls made by this method to <paramref name="input" />, may under certain conditions result in exceptions thrown by the stream.
+        /// </remarks>
+        public async Task UploadFileAsync(Stream input, string path, UploadMode createMode, CancellationToken cancellationToken)
```

Comment:
At first, I was inclined to say "use System.IO.FileMode directly, and remove the UploadMode enum".
But I'd actually prefer to remove this method altogether. I don't see an advantage over using OpenAsync(string path, FileMode mode, FileAccess access, CancellationToken cancellationToken) and having users invoke Stream.CopyToAsync(...) themselves. This gives them also control over the buffer size that is used to read from the input stream. Right now, you use our "optimal" write buffer size for this.

Reply:
I created the new enum just to be semantically explicit and for better developer experience.
I think that the use of optimal buffer size is the reason why we would have this method in the first place. As a developer without any knowledge of how the library works, I would probably choose my "go to" number 4096, which would lead to suboptimal performance in this case.
If I disregard buffer size selection, I agree that users could easily implement this themselves, but then the same argument could be had for the sync version ```UploadFile```.